### PR TITLE
View raster and mesh files from the Danesfield workflow

### PR DESCRIPTION
**Pull request by mvandenburgh**
We need a way to associate `Image`, `Mesh3D`, and `Tiles3D` objects generated from a `Dataset`'s output files with the `Dataset` itself.

@AlmightyYakob do you think this would be appropriate to upstream to OASIS? Then, we wouldn't need a separate model and could instead just put these `ManyToManyFields` directly on the `Dataset` model. I'm not certain that this would be in scope for OASIS though.

### CODE REVIEW

File: danesfield/core/models/dataset.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+class IngestedOutputFiles(models.Model):
+    dataset = models.ForeignKey(Dataset, on_delete=models.CASCADE, related_name='ingested_files')
+
+    imagery_files = models.ManyToManyField(Image, related_name='+')
+    mesh_3d_files = models.ManyToManyField(Mesh3D, related_name='+')
+    tiles_3d_files = models.ManyToManyField(Tiles3D, related_name='+')
+
+    @classmethod
+    @transaction.atomic
+    def from_dataset(cls, dataset: Dataset):
```
Comment (AlmightyYakob) [inline danesfield/core/models/dataset.py:22]: This function doesn't touch the `tiles_3d_files` field at all. Is that waiting on [this](https://github.com/ResonantGeoData/ResonantGeoData/pull/618) PR?

**Comment by AlmightyYakob** [comment]
Just so I'm clear, the main reason for needing this is so that given a dataset, the _ingested files_ can easily be accessed, not just the checksum files?

> @AlmightyYakob do you think this would be appropriate to upstream to OASIS? Then, we wouldn't need a separate model and could instead just put these `ManyToManyFields` directly on the `Dataset` model. I'm not certain that this would be in scope for OASIS though.

I'd like to see how it shakes out here first, and then evaluate. I'm not against it necessarily, after all OASIS is using the RGD ecosystem, but I think using it here first is a better idea than jumping to upstreaming it.

**Comment by mvandenburgh** [comment]
> Just so I'm clear, the main reason for needing this is so that given a dataset, the _ingested files_ can easily be accessed, not just the checksum files?

Right. Now that I think about it, there's also an alternative to this that doesn't require a new model.

```
def get_ingested_file(file: ChecksumFile):
	try:
		return Image.objects.get(file=file)
	except DoesNotExist:
		try:
			return Mesh3D.objects.get(file=file)
		except DoesNotExist:
			try:
				return Tiles3D.objects.get(file=file)
			except DoesNotExist:
				return None
```

This approach has the advantage of not requiring any database changes. WDYT @AlmightyYakob ? I think this might be a better approach, the downside is we'd be making 2 and 1 extra queries for Mesh3D and Tiles3D, respectively, but since that's a constant number of queries I don't think it's a big deal. 
		

> > @AlmightyYakob do you think this would be appropriate to upstream to OASIS? Then, we wouldn't need a separate model and could instead just put these `ManyToManyFields` directly on the `Dataset` model. I'm not certain that this would be in scope for OASIS though.
> 
> I'd like to see how it shakes out here first, and then evaluate. I'm not against it necessarily, after all OASIS is using the RGD ecosystem, but I think using it here first is a better idea than jumping to upstreaming it.

:+1:

File: danesfield/core/models/dataset.py (hunk 1 above)
Comment (mvandenburgh) [reply danesfield/core/models/dataset.py:22]: Technically yes it's waiting on that PR. I was able to pin the RGD django packages to that PR branch so we can use the code in the meantime, but I forgot to update this function - I'll go ahead and do that

**Comment by AlmightyYakob** [comment]
> Right. Now that I think about it, there's also an alternative to this that doesn't require a new model.
> 
> ```
> def get_ingested_file(file: ChecksumFile):
> 	try:
> 		return Image.objects.get(file=file)
> 	except DoesNotExist:
> 		try:
> 			return Mesh3D.objects.get(file=file)
> 		except DoesNotExist:
> 			try:
> 				return Tiles3D.objects.get(file=file)
> 			except DoesNotExist:
> 				return None
> ```
> 
> This approach has the advantage of not requiring any database changes. WDYT @AlmightyYakob ? I think this might be a better approach, the downside is we'd be making 2 and 1 extra queries for Mesh3D and Tiles3D, respectively, but since that's a constant number of queries I don't think it's a big deal.

I like this approach, We'd be making a max of 3 queries, and a min of one, which isn't too bad. I'd propose the following code suggestion though

```python
def get_ingested_file(file: ChecksumFile):
    for model in [Image, Mesh3D, Tiles3D]:
        try:
            return model.objects.get(file=file)
        except DoesNotExist:
            pass
    
    return None
```

This would be hacky, but we also could make use of the `ChecksumFile.description` field, to add a tag that indicates which model it uses. That might introduce other issues though, so for now I think this is fine.

**Comment by mvandenburgh** [comment]
> > Right. Now that I think about it, there's also an alternative to this that doesn't require a new model.
> > ```
> > def get_ingested_file(file: ChecksumFile):
> > 	try:
> > 		return Image.objects.get(file=file)
> > 	except DoesNotExist:
> > 		try:
> > 			return Mesh3D.objects.get(file=file)
> > 		except DoesNotExist:
> > 			try:
> > 				return Tiles3D.objects.get(file=file)
> > 			except DoesNotExist:
> > 				return None
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > This approach has the advantage of not requiring any database changes. WDYT @AlmightyYakob ? I think this might be a better approach, the downside is we'd be making 2 and 1 extra queries for Mesh3D and Tiles3D, respectively, but since that's a constant number of queries I don't think it's a big deal.
> 
> I like this approach, We'd be making a max of 3 queries, and a min of one, which isn't too bad. I'd propose the following code suggestion though
> 
> ```python
> def get_ingested_file(file: ChecksumFile):
>     for model in [Image, Mesh3D, Tiles3D]:
>         try:
>             return model.objects.get(file=file)
>         except DoesNotExist:
>             pass
>     
>     return None
> ```
> 
> This would be hacky, but we also could make use of the `ChecksumFile.description` field, to add a tag that indicates which model it uses. That might introduce other issues though, so for now I think this is fine.

Sounds good, I'll change this PR to do it this way instead

**Comment by mvandenburgh** [comment]
Peek 2022-01-12 17-12

Changes I've made:
- Add an additional endpoint that, given the name of a file in a `Dataset`, redirects the user to the RGD server-rendered page for the ingested version of that file (i.e. `Rasters` for 2D images and `Mesh3D` for 3D meshes, 3D tiles are not implemented yet).
- Add logic to the frontend to use this endpoint to render the RGD page as an iframe

Adding an additional endpoint that redirects to another page on the server seems a little hacky, but I couldn't think of a cleaner way to do it; I'm open to suggestions if you can think of a better way.

File: client/src/components/FileBrowser.vue [hunk 2]
```diff
@@ -53,6 +56,17 @@
               mdi-download
             </v-icon>
           </v-btn>
+          <v-btn
+            icon
+            @click="setFile(item)"
```
Comment (AlmightyYakob) [inline client/src/components/FileBrowser.vue:61]: Should this be `setOpenFile`? I can't fine `setFile` anywhere else

```suggestion
            @click="setOpenFile(item)"
```

**Comment by AlmightyYakob** [comment]
> Adding an additional endpoint that redirects to another page on the server seems a little hacky, but I couldn't think of a cleaner way to do it; I'm open to suggestions if you can think of a better way.

I don't think it's hacky, as that endpoint takes something known to us as a single type (a file in a dataset), and returns various things based on inspecting that file (which we have no real idea about from the client). Since we're dealing with files that map to various models, I think it's fine.

**Comment by AlmightyYakob** [comment]
@mvandenburgh I guess we can just merge this PR now and create an issue for adding in 3d tiles support?

**Comment by mvandenburgh** [comment]
> @mvandenburgh I guess we can just merge this PR now and create an issue for adding in 3d tiles support?

Sounds good to me.

**Pull request merged**